R/util_data_summaries.R

```diff
@@ -126,3 +155,509 @@ ds.summarize_by_column <- function(
+        if( is.null(cov_vec) | identical(cov_vec,"") ){
+            # there is an empty (unadjusted) cov_group
+            return("")
+        } else{
+            return( " + " %+% paste(cov_vec, collapse=" + ") )
+        }
+    }) %>% unlist()
+    
+    # build the formula data frame (fdf) as each combination of
+    # variable strings
+    fdf <- expand.grid(
+        cov_str = cov_groups_strs,
+        mod_str = mod_strs,
+        iv_str  = iv_strs,
+        dv_str  = dv_strs
+    )
+    
+    # assemble the formula grid into 
+    formula_vec <- fdf$dv_str %+% fdf$iv_str %+% fdf$mod_str %+% fdf$cov_str
+    return(formula_vec)
+    
+}
+
+ds.helper$glm1_formula_to_varlist <- function(formula){
+    # input is a canonical ds.glm1 formula, 
+    # see ds.validate_glm1_model
+    # ouput is ds-style glm1 model variable list
+    
+    varlist <- list( dv=NA, iv=NA, mod=NA, covs=NA )
+    split_on_tilde <- str_split(formula,"[ ]*~[ ]*")[[1]]
+    
+    # check if formula has dvs and predictors
+    if( length(split_on_tilde) != 2){
+        stop("invalid ds.glm1 formula provided: " %+% formula)
+    } else if(nchar(split_on_tilde[1]) == 0 | 
+              nchar(split_on_tilde[2]) == 0 ){
+        stop("invalid ds.glm1 formula provided: " %+% formula)
+    }
+    
+    varlist$dv <- split_on_tilde[1]
+    predictor_str <- split_on_tilde[2]
+    
+    # this block finds the iv and moderator
+    # there are moderators if there is a star and
+    # it comes before any pluses, if pluses are present
+    split_at_star <- str_split(predictor_str,"[ ]*\\*[ ]*")[[1]]
+    if( length(split_at_star) > 1){
+        # there might be a moderator (because there's a star)
+        # but it might also be a non-canonical ds.glm formula
+        pre_star  <- split_at_star[1]
+        post_star <- split_at_star[2]
+        if( length(split_at_star) > 2 | grepl( "\\+", pre_star) ){
+            "ds.glm1 supports [0,1] stars" %+%
+                " and only between the 1st and 2nd predictor," %+%
+                " but this formula was provided: " %+% formula %>%
+                stop()
+        }
+        # mod is 1st term post_star and before + or end
+        varlist$iv  <- pre_star
+        varlist$mod <- str_split(post_star,"[ ]*\\+[ ]*")[[1]][1]
+    } else{
+        # there was no moderator, so the iv is the first predictor
+        # before a plus or before the end of the predictors string
+        varlist$iv <- str_split(predictor_str,"[ ]*\\+[ ]*")[[1]][1]
+    }
+    
+    # find any covariate(s)
+    match_plus_or_star <- "[ ]*\\+[ ]*|[ ]*\\*[ ]*"
+    predictors <- str_split( predictor_str, match_plus_or_star)[[1]]
+    if( sum( ! predictors %in% varlist ) > 0){
+        varlist$covs <- predictors[ ! predictors %in% varlist ]
+    }
+    
+    return(varlist)
+}
+
+
+ds.glm1 <- function( formula, data, mod_family=NULL ){
+    # run the appropriate model, unless mod_family override is supplied
```

Comment:
nit: "mod_family" sounds like it could mean "moderator_family"—I'd just call it "model_family" for clarity.